### ISSUE TITLE
Implement hierarchical clustering

### ISSUE THREAD

**Issue by Mec-iS** [smartcorelib/smartcore#11]
Motivation: why do we need hierarchical when we have already kmeans?

Vocabulary:
* divisive clustering: ...
* agglomerative clustering: average, weighted, median, centroid, Ward

Sub-tasks:
* pick one or a minimal set of metrics-distances
* pick one or a minimal set of linkage strategies
* pick one or more [algorithms](https://en.wikipedia.org/wiki/Hierarchical_clustering#Open_source_implementations) (SLINK for single-linkage and CLINK for complete-linkage clustering)

Visualisations: (?)

Other implementations:
* hierarchical clustering [in SciPy](https://docs.scipy.org/doc/scipy/reference/cluster.hierarchy.html#module-scipy.cluster.hierarchy)
* hierarchical clustering [in sklearn](https://scikit-learn.org/stable/modules/clustering.html#hierarchical-clustering)

**Comment by VolodymyrOrlov** [comment]
@Mec-iS Take a look at [this paper](https://arxiv.org/pdf/cs/9912014.pdf) that describes a FastPair algorithm. This algorithm helps to speedup cluster merge operation.

**Comment by Mec-iS** [comment]
## Notes

### FastPair
* a [Python implementation for FastPair](https://github.com/carsonfarmer/fastpair/blob/master/fastpair/base.py)
* [Closest neighbour on Rosetta Stone](http://rosettacode.org/wiki/Closest-pair_problem#Python), see the usage of complex numbers (for 2D points) to reduce distance to absolute value of difference (`d = abs(P1-P2)`)
* Closest neighbout using randomness: [Rabin’s Algorithm](https://rjlipton.wordpress.com/2009/03/01/rabin-flips-a-coin/)

### `fastcluster`:
* Direct link to [download the Paper](https://github.com/dmuellner/fastcluster/raw/master/docs/fastcluster.pdf) with extensive explanation of the procedure
* [Python interface](https://github.com/dmuellner/fastcluster/blob/master/fastcluster.py) for version 1.1.11 (from 2013), current version is 1.1.26 (December 2019) with a [C++](https://github.com/dmuellner/fastcluster/blob/master/src/fastcluster.cpp) and a [CPython](https://github.com/dmuellner/fastcluster/blob/master/src/fastcluster_python.cpp) version

#### Alternatives:
1. translate from the Python interface in Rust, then hunt for changes/improvements in the C++ version
2. FFI to call C++ from Rust, in particular using [`rustcxx`](https://github.com/google/rustcxx)

Background questions:
* are we going for a 100% Rust native implementation?
* are we supposed to allow or not usages of `unsafe` blocks?

**Comment by VolodymyrOrlov** [comment]
> Background questions:
> 
> * are we going for a 100% Rust native implementation?
> * are we supposed to allow or not usages of `unsafe` blocks?

Yes, calling C++ library from SmartCore is not an option for multiple reasons. We'll have to ship `fastcluster` with SmartCore somehow and it diminishes usefulness of our library. 

Do you know C++ by any chance? 😄  If not, feel free to go with any implementation, even if it is not the fastest out there. Another option would be to try to implement the algorithm ([it is described here](https://arxiv.org/pdf/1109.2378.pdf)) yourself. It would be super awesome if you can implement `fastcluster` in Rust, because in this case we will be the only library in Rust that has it.

**Comment by Mec-iS** [comment]
what role you had in mind for FastPair?

**Comment by VolodymyrOrlov** [comment]
> what role you had in mind for FastPair?

As an alternative to `fastcluster`

**Comment by Mec-iS** [comment]
#142
